### PR TITLE
Compatibility with Symfony3 for 1.3.x

### PR DESCRIPTION
**Pull request by covex-nn**
Unit tests are passed, users can register, confirm registration, login, logout, view and adit their profiles, change or reset password. All commands can be executed from cli.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,15 +1,15 @@
 language: php
 
 php:
-  - 5.3
-  - 5.4
```
Comment (xabbuh) [inline .travis.yml]: Why did you remove these lines? FOSUserBundle 1.3 dies work with PHP 5.3 and so tests should be executed on this version too.

File: composer.json [hunk 2]
```diff
@@ -21,16 +21,18 @@
     ],
     "require": {
         "php": ">=5.3.2",
-        "symfony/framework-bundle": "~2.3",
-        "symfony/security-bundle": "~2.3"
+        "symfony/framework-bundle": "~2.6|~3.0",
+        "symfony/http-kernel": "~2.6|~3.0",
+        "symfony/security-bundle": "~2.6|~3.0",
+        "symfony/form": "~2.6|~3.0"
```
Comment (xabbuh) [inline composer.json]: Forbidding Symfony versions before 2.6 is a BC break that cannot be done in a minor version release (not to talk about patch releases).

File: .travis.yml (hunk 1 above)
Comment (covex-nn) [inline .travis.yml]: There was no reason. I was wrong. I thought about Symfony3 and its requirements.

Comment (covex-nn) [inline .travis.yml]: No! I remembered! There was a reason: i used **Symfony\Component\DependencyInjection\ContainerAwareTrait** in controllers because in Symfony3 **Symfony\Component\DependencyInjection\ContainerAware** was removed and there was no traits in PHP 5.3 and PHP 5.4

Comment (xabbuh) [inline .travis.yml]: Well, this means that you cannot use the trait in a bundle that needs to maintain compatibility with PHP 5.3.

File: composer.json (hunk 2 above)
Comment (covex-nn) [inline composer.json]: I will try to fix code to use Symfony components ~2.3 and ~3.0 and i will test everything with Symfony SE >= 2.3

File: .travis.yml (hunk 1 above)
Comment (covex-nn) [inline .travis.yml]: You are right. There was no need in traits. I will fix it. All controllers will extend new abstract class FOS\UserBundle\Controller\ContainerAwareController

**Comment by Pierstoval** [comment]
For Symfony3, use the `dev-master` branch. I think it's better to ask for a new FOSUserBundle tag than make 1.3 compatible with SF3, because there are so many differences and there can be many BC breaks :confused:

**Comment by covex-nn** [comment]
I cannot use `dev-master` =(
And there are only few things to fix back and to make compatible back with `symfony components ~2.3` and `php 5.3`! =)

* http://symfony.com/blog/new-in-symfony-2-4-the-request-stack
* http://symfony.com/blog/new-in-symfony-2-6-security-component-improvements
* http://symfony.com/doc/2.5/components/console/helpers/questionhelper.html
* And some of my changes when i hoped that i can change dependency to `~2.6 | ~3.0`

**Comment by Pierstoval** [comment]
If you want to update 1.3 branch, according to semver, you **must** keep backward compatibility with symfony 2.3, so for the request stack example you have to create different definitions depending on versions, etc.
I don't think it's worth it as 2.x (master) branch is already compatible and just needs a new alpha release (or a stable release, because I don't see bugs in master actually, but it's up to the maintainers).

**Comment by covex-nn** [comment]
@xabbuh @Pierstoval I fixed code. Now bundle is compatible with Symfony 2.3 and with Symfony 3.0

**Comment by Pierstoval** [comment]
:+1:

File: .travis.yml [hunk 3]
```diff
@@ -6,10 +6,34 @@ php:
...
+    - php: '5.4'
+      env: SYMFONY_VERSION=2.3.*
+    - php: '5.5'
+      env: SYMFONY_VERSION=2.3.*
+    - php: '5.6'
+      env: SYMFONY_VERSION=2.3.*
+    - php: 'hhvm'
+      env: SYMFONY_VERSION=2.3.*
+    - php: '7.0'
+      env: SYMFONY_VERSION=2.3.*
```
Comment (Pierstoval) [inline .travis.yml]: You should test on 2.8, 3.0 and 3.1 too

Comment (covex-nn) [inline .travis.yml]: Done

* https://travis-ci.org/FriendsOfSymfony/FOSUserBundle/builds/144110094
* on php 5.3/5.4 bundle tested on **2.3** and **~2.3 | ~3.0** (i.e. 2.8)
* on php 5.5/5.6/hhvm/7.0 bundle tested on **2.3**, **2.8**, **3.0** and **~2.3 | ~3.0** (i.e. 3.1)

**Comment by stof** [comment]
@covex-nn please disable Travis on your fork. The .travis.yml file is configured to send notifications to the FOS mailing-list so FOS admins are receiving emails asking to moderate the Travis notifications each time you push a commit in your fork. Tests are already running when Travis tests the PR on the main repo

File: .travis.yml [hunk 4]
```diff
@@ -6,10 +6,50 @@ php:
...
+    - php: 'hhvm'
+      env: SYMFONY_VERSION=2.8.*
+    - php: 'hhvm'
+      env: SYMFONY_VERSION=3.0.*
+    - php: '7.0'
+      env: SYMFONY_VERSION=2.3.*
+    - php: '7.0'
+      env: SYMFONY_VERSION=2.8.*
+    - php: '7.0'
+      env: SYMFONY_VERSION=3.0.*
```
Comment (stof) [inline .travis.yml]: Testing all combinations of Symfony and PHP is a bad idea. This makes the build matrix explode, and we are only allowed to have 5 concurrent jobs on Travis for the whole FOS organization. So creating 20 jobs for each build of FOSUserBundle is a no-go.

File: Controller/SecurityController.php [hunk 5]
```diff
@@ -38,14 +46,11 @@ public function loginAction()
...
         // last username entered by the user
-        $lastUsername = (null === $session) ? '' : $session->get(SecurityContext::LAST_USERNAME);
-
-        $csrfToken = $this->container->has('form.csrf_provider') ? $this->container->get('form.csrf_provider')->generateCsrfToken('authenticate') : null;
+        $lastUsername = (null === $session) ? '' : $session->get($lastUsername);
 
         return $this->renderLogin(array(
             'last_username' => $lastUsername,
             'error'         => $error,
-            'csrf_token' => $csrfToken,
```
Comment (stof) [inline Controller/SecurityController.php]: You are removing features here

File: DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -109,6 +114,11 @@ private function addProfileSection(ArrayNodeDefinition $node)
 
     private function addRegistrationSection(ArrayNodeDefinition $node)
     {
+        if (method_exists('Symfony\Component\Form\FormTypeInterface', 'setDefaultOptions')) {
```
Comment (stof) [inline DependencyInjection/Configuration.php]: This is broken, because it will use the deprecated API on Symfony 2.8, where both system exist...

File: Form/Handler/ChangePasswordFormHandler.php [hunk 7]
```diff
@@ -15,18 +15,15 @@
...
 
-class ChangePasswordFormHandler
+class ChangePasswordFormHandler extends RequestFormHandler
 {
-    protected $request;
     protected $userManager;
     protected $form;
 
-    public function __construct(FormInterface $form, Request $request, UserManagerInterface $userManager)
+    public function __construct(FormInterface $form, UserManagerInterface $userManager)
```
Comment (stof) [inline Form/Handler/ChangePasswordFormHandler.php]: Changing the constructor is a BC break. It cannot be done in 1.3.x

File: Form/Handler/ChangePasswordFormHandler.php [hunk 8]
```diff
@@ -15,18 +15,15 @@
 use FOS\UserBundle\Model\UserInterface;
 use FOS\UserBundle\Model\UserManagerInterface;
 use Symfony\Component\Form\FormInterface;
-use Symfony\Component\HttpFoundation\Request;
 
-class ChangePasswordFormHandler
+class ChangePasswordFormHandler extends RequestFormHandler
 {
-    protected $request;
```
Comment (stof) [inline Form/Handler/ChangePasswordFormHandler.php]: Removing protected properties is also a BC break

File: Resources/config/change_password.xml [hunk 9]
```diff
@@ -19,10 +20,15 @@
...
-        <service id="fos_user.change_password.form.handler.default" class="FOS\UserBundle\Form\Handler\ChangePasswordFormHandler" scope="request">
+        <service id="fos_user.change_password.form.handler.default" class="FOS\UserBundle\Form\Handler\ChangePasswordFormHandler">
             <argument type="service" id="fos_user.change_password.form" />
-            <argument type="service" id="request" />
             <argument type="service" id="fos_user.user_manager" />
+            <call method="setRequestStack">
+                <argument type="service" id="request_stack" on-invalid="ignore" />
+            </call>
+            <call method="setRequest">
+                <argument type="service" id="request" on-invalid="ignore" />
```
Comment (stof) [inline Resources/config/change_password.xml]: This will still trigger deprecations in Symfony 2.8

**Comment by stof** [comment]
There are multiple BC breaks in this PR so it cannot be merged into 1.3.x (and some cases are impossible to upgrade without BC breaks btw).

I will work on releasing FOSUserBundle 2.x in the next few days. I don't think it is worth upgrading 1.3.x, given the amount of work necessary to ensure we don't break BC (I commented on some BC breaks in your PR, but there are many more places breaking BC actually)

**Comment by covex-nn** [comment]
@stof ok, i understood. I cannot keep and cannot remove `Request $request` from `__construct` parameters list, it is imposible to fix and it cannot be merged into `1.3.x` =(

Anyway the main goal was to have a stable version of FOSUserBundle for Symfony3. I can reach it with stable 2.x too =) Thank you
